contracts/utils/structs/MerkleTree.sol

```diff
@@ -0,0 +1,189 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import {MerkleProof} from "../cryptography/MerkleProof.sol";
+import {Arrays} from "../Arrays.sol";
+import {Panic} from "../Panic.sol";
+
+/**
+ * @dev A complete binary tree with the ability to sequentially insert leaves, changing them from a zero to a non-zero
+ * value, while keeping a history of merkle roots. This structure allows inserting commitment (or other entries) that
+ * are not stored, but can be proven to be part of the tree.
+ *
+ * The history of merkle roots allow inclusion proofs to remain valid even if leaves are inserted into the tree between
+ *  the moment the proof is generated and the moment it's verified.
+ *
+ * Each tree can be customized to use specific
+ * - depth
+ * - length of the root history
+ * - zero values (for "empty" leaves)
+ * - hash function
+ *
+ * WARNING:
+ *
+ * By design, the tree include zero leaves. Customizing the "zero value" might be necessary to ensure that empty leaves
+ * being provably part of the tree is not a security issue.
+ *
+ * _Available since v4.x._
+ */
+library MerkleTree {
+    /**
+     * @dev Maximum supported depth. Beyond that, some checks will fail to properly work.
+     * This should be enough for any realistic usecase.
+     */
+    uint256 private constant MAX_DEPTH = 255;
+
+    /**
+     * @dev Merkle tree cannot be initialized because requested depth is to large.
+     */
+    error MerkleTreeInvalidDepth(uint256 depth, uint256 maxDepth);
+
+    /**
+     * @dev The `sides` and `zero` arrays are set, at initialization, to have a length equal to the depth of the tree.
+     * No push/pop operations should be performed of these array, and their lengths should not be updated.
+     *
+     * The `roots` array stores the history of roots. Its length is set at initialization, and should not be updated.
+     *
+     * The hashing function used during initialization to compute the `zeros` values (value of a node at a given depth
+     * for which the subtree is full of zero leaves). This function is kept in the structure for handling insertions.
+     */
+    struct TreeWithHistory {
+        uint256 currentRootIndex;
+        uint256 nextLeafIndex;
+        bytes32[] sides;
+        bytes32[] zeros;
+        bytes32[] roots;
+        function(bytes32, bytes32) view returns (bytes32) fnHash;
```

Comment:
If we receive a `function(bytes)` instead, we can add built-in leaf assymetry, I think we should strive to have assymetry before releasing this tree.

Reply:
I'm not sure what bytes would bring here ... appart from the cost of encoding/decoding.

Reply:
We would be able to do `fnHash(fnHash(...))` for the leafs 🤔

Reply:
I don't think this structure should take care of hashing the leafs. The dev using that should hash the leafs, then inject them into that structure, and we only take care of maintaining internal nodes. We don't know which hashes are going to be used in the future, both for the leaf part and the internal part. Particularly in the case of ZK structure. 
I don't want to make the (strong) assumption that leafs are generated by applying twice the internal hash.

Reply:
I don't want to get into trouble, so I'm not going to post T*rnad* code here.
I can share it privatly on slack.

Reply:
Right, sorry for bringing the assymetry discussion back again haha

> The dev using that should hash the leafs, then inject them into that structure, and we only take care of maintaining internal nodes. 

My feeling is that we need an usage example in the docs and also make clear this is the intention. Otherwise the convo will come back again.

Reply:
Solving and adding this to the docs backlog